Fix #315 #342

pvillega opened the pull request:
Fix issue in Bits where singleBit for Long was shitfing bits incorrectly

Add random index for IndexTests/OptionalTests using ScalaCheck

julien-truffaut commented on build.sbt:
```diff
@@ -43,6 +43,7 @@ lazy val buildSettings = Seq(
 lazy val scalaz     = Def.setting("org.scalaz"      %%% "scalaz-core" % "7.2.1")
 lazy val shapeless  = Def.setting("com.chuusai"     %%% "shapeless"   % "2.3.0")
 lazy val refinedDep = Def.setting("eu.timepit"      %%% "refined"     % "0.3.7")
+lazy val refinedScalaCheckDep = Def.setting("eu.timepit"      %%% "refined-scalacheck"  % "0.3.7")
```
can you add "test" modifier

pvillega commented:
Updated! I had to move the dependency for it to compile, let me know if there's anything else you want changed :)

julien-truffaut commented on law/shared/src/main/scala/monocle/law/discipline/function/AtTests.scala:
```diff
@@ -7,18 +7,17 @@ import org.scalacheck.Arbitrary
+    new SimpleRuleSet("At", LensTests(at(_: I)).props: _*)
+  }
 
   def defaultIntIndex[S, A](implicit aEq: Equal[A], aArb: Arbitrary[A],
```
we can probably remove `defaultIntIndex`

julien-truffaut commented on law/shared/src/main/scala/monocle/law/discipline/function/AtTests.scala:
```diff
@@ -7,18 +7,17 @@ import org.scalacheck.Arbitrary
-                                   evAt: At[S, I, A]): RuleSet =
-    new SimpleRuleSet("At", LensTests(at(i)).props: _*)
+  def apply[S, I, A](implicit aEq: Equal[A], aArb: Arbitrary[A],
+                     sEq: Equal[S], sArb: Arbitrary[S],
```
do you mind to maintain the vertical alignment

julien-truffaut commented on law/shared/src/main/scala/monocle/law/discipline/function/IndexTests.scala:
```diff
@@ -10,14 +10,14 @@ import scalaz.Equal
+                     iArb: Arbitrary[I], evIndex: Index[S, I, A]): RuleSet =
+    new SimpleRuleSet("Index", OptionalTests(index(_ : I)).props: _*)
 
   def defaultIntIndex[S, A](implicit aEq: Equal[A], aArb: Arbitrary[A],
```
same

julien-truffaut commented:
Thank you very much, it is awesome to write tests that actually find bugs :)

pvillega commented:
No, thank you! It was fun and I learned :) 

I've removed the methods, updated code to user the 'proper' apply way, and fixed the styling (hopefully). Let me know if you want me to tweak something else, as IntelliJ may do stuff I don't realise... :eh:

julien-truffaut commented:
:+1:
